### PR TITLE
fix(transcript_logger): consecutive-duplicate turn guard (#968)

### LINKED ISSUE #968: transcript_logger: identical turns appended Nx under duplicated hook registration — add consecutive-duplicate guard
## Observed

In a captured local project transcript (`<git-common-dir>/aelfrice/transcripts/turns.jsonl`), one session dated 2026-05-26 has **every turn appended exactly 4×** — identical `role`/`text`/`session_id`, distinct `turn_id`s, all four writes within ~20 ms. Other sessions in the same file (before and after) are clean, which points at a per-environment cause — most plausibly duplicated UserPromptSubmit/Stop hook registration (e.g., stacked settings in a worktree) — rather than a steady-state logger bug.

## Why it matters

`turns.jsonl` is treated as ground truth downstream: the PreCompact rebuilder's recent-turns window, cadence turn-density scoring, and eval-fixture extraction all silently inflate when a session's turns are duplicated. A 4×-duplicated session reads as 4× the conversational density it actually had.

## Current behavior

The append path in `src/aelfrice/transcript_logger.py` is append-only under the non-blocking contract, with no idempotency guard — N hook invocations for one event produce N identical lines.

## Proposal

Consecutive-duplicate guard at append time: skip the write when the file's last line has the same `(session_id, role, text)` and a `ts` within a short window (~2 s); count skips (e.g., `skipped_duplicate` in the hook-audit row). Preserves the non-blocking contract; no new locking.

## Acceptance
- A 4×-burst of identical hook fires produces one transcript line.
- Skip count is observable in `hook_audit.jsonl`.
- Test covering the burst shape (identical text/role/session, distinct turn_ids, sub-second spacing).
- Distinct-text appends within the window are unaffected.

### PR DESCRIPTION
**Pull request by robotrocketscience**
Closes #968.

## Problem
A duplicated hook registration (stacked settings in a worktree) fires `UserPromptSubmit`/`Stop` N times for one event, so `transcript_logger` appends N identical turn lines ~20ms apart (same `role`/`text`/`session_id`, distinct `turn_id`s). Because `turns.jsonl` is ground truth downstream — the PreCompact rebuilder's recent-turns window, cadence turn-density scoring, eval-fixture extraction — a 4x-duplicated session silently reads as 4x its real density.

## Fix
Consecutive-duplicate guard at append time:
- A turn whose `(session_id, role, text)` matches the file's previous turn line within `DUP_WINDOW_SECONDS` (2s) is dropped.
- The drop is recorded as a `skipped_duplicate` row in `hook_audit.jsonl` (observability).
- The previous line is found via a bounded tail read (`_TAIL_READ_BYTES`), keeping the per-append cost O(1) in file size and inside the sub-10ms budget. No new locking — the non-blocking contract holds.
- Compaction markers append directly and are exempt.
- Fails open: any record not positively confirmed as a duplicate is appended, so distinct text and deliberate resends past the window are unaffected.

## Supporting refactor (separate commit)
Writing the skip to `hook_audit.jsonl` required reaching the audit sink, but importing `aelfrice.hook` pulls in the retrieval stack (scipy, ~220ms cold) — fatal on a sub-10ms hook path. The first commit extracts the scipy-free audit primitives (`HookAuditConfig`, `load_hook_audit_config`, `_audit_path_for_db`, `_append_audit`, constants) into a new `aelfrice.hook_audit` module. `hook.py` re-exports them (no caller breaks); `hook_tail` imports from the new module directly. Behavior-preserving.

## Acceptance (from #968)
- [x] A 4x-burst of identical hook fires produces one transcript line.
- [x] Skip count is observable in `hook_audit.jsonl`.
- [x] Test covers the burst shape (identical text/role/session, distinct turn_ids, sub-second spacing).
- [x] Distinct-text appends within the window are unaffected.

## Verification
- `tests/test_transcript_logger.py` (+9 new tests), `tests/test_hook_audit.py`, `tests/test_hook_tail.py`, `tests/test_hook_user_prompt_submit.py`, `tests/test_hook_import_resilience.py`: 150 passed.
- `aelfrice.hook_audit` / `aelfrice.transcript_logger` confirmed scipy-free (8–36ms import).

## Summary by Sourcery

Add a consecutive-duplicate guard to transcript logging to collapse bursty duplicate turns while recording skips in the hook audit log, and extract audit primitives into a lightweight module for reuse off the heavy retrieval import path.

Enhancements:
- Introduce a time-windowed consecutive-duplicate filter for user and assistant turns in transcript_logger, preserving lock-free writes and exempting compaction markers.
- Record skipped duplicate turns as structured entries in the hook audit log for observability, using lazy access to the existing audit sink.
- Extract per-turn audit configuration, path resolution, and append/rotate helpers from hook.py into a new aelfrice.hook_audit module and re-export them for backward compatibility.

Tests:
- Add tests covering burst deduplication behavior, window-bound duplicate handling, compaction marker interactions, and audit logging of skipped duplicates.

## Release Notes

* **New Features**
  * Consecutive duplicate detection: Identical message submissions within a time window are automatically detected and deduplicated, preventing accidental re-submissions from being logged.
  * Audit visibility: Skipped duplicate turns are recorded in the audit log for tracking and observability.

* **Tests**
  * Added comprehensive test suite validating duplicate detection across various scenarios.

### CODE REVIEW

**Comment by robotrocketscience** [comment]
[claim:review:Kulili:2026-06-18T23:57:40Z]

File: src/aelfrice/transcript_logger.py [hunk 1]
```diff
@@ -129,6 +140,135 @@
...
+            return
+        record: dict[str, object] = {
+            "ts": _now_iso(),
+            "hook": "transcript_logger",
+            "event": "skipped_duplicate",
+            "role": role,
+            "session_id": session_id,
+        }
+        _append_audit(_audit_path_for_db(p), record, cfg.max_bytes)
+    except Exception:
```
[the remark this answers is not in the document]
Comment (robotrocketscience) [reply src/aelfrice/transcript_logger.py:244]: Fixed in 9cb66dbe: added an inline comment on the except clause stating the swallow is intentional (audit is observability-only and must not break the non-blocking hook path).

File: src/aelfrice/hook.py [hunk 2]
```diff
@@ -40,6 +40,18 @@
...
+        HookAuditConfig,
+        _append_audit,
+        _audit_path_for_db,
+        _CONFIG_FILENAME,
+        load_hook_audit_config,
+    )
+    # Re-exported so existing `from aelfrice.hook import ...` callers keep
+    # working after the #968 extraction into aelfrice.hook_audit.
+    from aelfrice.hook_audit import AUDIT_DEFAULT_MAX_BYTES  # noqa: F401
+    from aelfrice.hook_audit import AUDIT_FILENAME  # noqa: F401
```
[the remark this answers is not in the document]
Comment (robotrocketscience) [reply src/aelfrice/hook.py:54]: Intentional re-export, not dead code: `AUDIT_FILENAME` was a public module constant on `aelfrice.hook` before this PR moved it to `hook_audit`, so the re-export preserves `from aelfrice.hook import AUDIT_FILENAME` for external (PyPI) callers. Same backward-compat re-export pattern cli.py uses (`# noqa: F401 — re-exported`). The `analyze (python)` gate passed; this is the known intentional-re-export false positive. Resolving.

**Comment by robotrocketscience** [comment]
Reviewed the diff. Approving (rebased onto current main + two CodeQL threads addressed).

**Correctness checks that pass:**
- **Dedup logic is sound.** Match key `(session_id, role, text)` correctly excludes `turn_id` (the burst carries distinct turn_ids). `abs()` on the ts delta tolerates clock skew. Compaction markers (no role/text) and any record with an unparseable `ts` compare unequal, so they never trigger a skip. A duplicate against a previously-skipped line still dedups because skipped writes don't advance the file tail — the whole burst collapses to one line.
- **Bounded, fail-open tail read.** `_read_last_turn_line` scans only the final 64 KiB (O(1) in file size) and drops the partial leading segment when the window starts mid-file, so a truncated record is never mistaken for a complete one; any parse failure yields `None` → append (never drops a turn it can't positively confirm as a duplicate).
- **Non-blocking contract preserved.** No new locking; the audit write is lazily imported and fully fail-soft.
- **Refactor is behavior-preserving.** `hook_audit` extraction keeps `hook.py` re-exports (`# noqa: F401`), matching the existing `cli.py` re-export convention; `hook_tail` imports from the new scipy-free module directly.

**Verification:** affected suites (transcript_logger / hook_audit / hook_tail / hook_user_prompt_submit / hook_import_resilience) = 150 passed locally. Discretion grep clean.

**Review-thread fixups (this branch, post-rebase):**
- `9cb66dbe` — added an explanatory comment on the fail-soft `except` in `_record_skipped_duplicate` (CodeQL empty-except finding).
- The `AUDIT_FILENAME` "unused import" is the intentional public-API re-export (preserves `from aelfrice.hook import AUDIT_FILENAME` for PyPI callers); resolved with rationale, not deleted.
